### ISSUE TITLE
HSTS cannot be disabled and should not be enforced if HTTP is on

### ISSUE THREAD

**Issue by raffaeler** [keycloak/keycloak#14162]
### Describe the bug

I disabled HSTS from the dockerfile with:

```
ENV KC_HOSTNAME_STRICT_HTTPS=false
```

When the container starts, I see this in the log:

```
2022-09-01 19:36:44,991 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: 
   FrontEnd: <request>,
   Strict HTTPS: false, Path: <request>,
   Strict BackChannel: false,
   Admin: <request>,
   Port: -1,
   Proxied: false
```

Anyway, after visiting the admin console, the chrome page chrome://net-internals/#hsts tells me that HSTS is enabled for localhost (yes, I published KC on localhost in container to test a bunch of scenarios).

In addition to that, you should also consider a scenario where HTTP is enabled with:
```
ENV KC_HTTP_ENABLED=true
```

In this case, enforcing HSTS is wrong because you'll end up in not being able to reach the HTTP site after visiting the HTTPS page once. In other words, if HTTP is enabled, HSTS should be automatically be turned off even without disabling the strict mode.

Conclusion:
1. `KC_HOSTNAME_STRICT_HTTPS=false` does not work
2. `KC_HTTP_ENABLED=true` should automatically set `KC_HOSTNAME_STRICT_HTTPS=false`

### Version

19.0.1

### Expected behavior

1. There should be a way to disable HSTS
2. When enforcing HTTP (which is correctly disabled by default), HSTS should be turned off

### How to Reproduce?

HSTS is always on regardless the settings are correctly set on Keycloak.
The user cannot visit any HTTP page, even if the pages belong to other apps and use different ports

**Comment by vmuzikar** [comment]
@raffaeler Thanks for the report and sorry for a late reply. `hostname-strict-https` is one of our internal options (hence not officially supported) and it is actually not setting HSTS. It is just affecting if (by default) Keycloak generated URLs must use HTTPS. You shouldn't need to override this option.

**Comment by raffaeler** [comment]
> @raffaeler Thanks for the report and sorry for a late reply. `hostname-strict-https` is one of our internal options (hence not officially supported) and it is actually not setting HSTS. It is just affecting if (by default) Keycloak generated URLs must use HTTPS. You shouldn't need to override this option.

Please remove it at least on `localhost`. It creates huge problems when testing (and during talks in conferences).

**Comment by JuryA** [comment]
I apologize for any confusion. Could you please confirm if HSTS is enabled by default in Keycloak, alongside HTTPS? Previous communication indicated that HTTPS is enabled by default, but it did not clarify the status of HSTS. Could you please provide some clarification? Thank you.

**Comment by raffaeler** [comment]
It is enabled by default and cannot be disabled, not even on localhost, which is a true mess.

**Comment by JuryA** [comment]
> It is enabled by default and cannot be disabled, not even on localhost, which is a true mess.

I understand your frustration… Given the inherent feature where a client (typically a browser) memorizes HSTS settings once the server announces its usage, it’s important to note that this setting cannot be easily reversed. Once the hostname is “tagged” for HSTS, it must continue using it indefinitely, or until the information is purged from the client’s HSTS cache. This might never happen, or if it does, it could take a year or more. While this is not a flaw but rather a deliberate security measure, it poses challenges during testing. I completely understand that Keycloak, being a crucial element in the security infrastructure, employs this feature. However, it’s unfortunate that there’s no way to disable it, even in a “non-production” mode. Is the a concern that it might be incorrectly implemented in a production environment?

**Comment by raffaeler** [comment]
I have no clue, but your guess is probably correct.
BTW I totally share the security concerns and agree that a production environment should always have HSTS turned on.

But there are cases where this need to be disabled. For example: localhost environment, test networks where you need to deploy multiple test services on the same domain.
I had to install a dns server to avoid messing the hosts file and demo the keycloak functionalities properly in my conference talks.

If I understand correctly, they simply do not care about scenarios like mine, that's all.

**Comment by vmuzikar** [comment]
The HSTS header ([`Strict-Transport-Security`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security)) is unrelated to the `hostname-strict-https` option. The header is configured on Realm basis in the Admin Console in the Security defenses tab.

**Comment by JuryA** [comment]
> The HSTS header ([`Strict-Transport-Security`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security)) is unrelated to the `hostname-strict-https` option. The header is configured on Realm basis in the Admin Console in the Security defenses tab.

Aha! Okay... Thank you for explanation! Now it´s clear to me... ;)

**Comment by demoncodemonkey** [comment]
**TLDR** to disable the pesky HSTS completely:
Keycloak -> Realm Settings -> Security Defenses tab -> HTTP Strict Transport Security (HSTS) -> `max-age=0`

**Comment by DamianEdwards** [comment]
Having HSTS default to on for new realms including when hosting on localhost is pretty developer hostile. HSTS should never be enabled for localhost. Ideally Keycloak would simply not set this header when the host is localhost.
